File: backend/controllers/class_controller.py
from datetime import datetime
from flask import request, jsonify, g
from controllers import handle_errors
import services.class_service as svc


def _parse_dt(s):
    if not s:
        return None
    for fmt in ("%Y-%m-%dT%H:%M", "%Y-%m-%dT%H:%M:%S", "%Y-%m-%d %H:%M:%S", "%Y-%m-%d %H:%M"):
        try:
            return datetime.strptime(s, fmt)
        except ValueError:
            pass
    return None


@handle_errors
def get_teacher_classes():
    return jsonify({"success": True, "data": svc.get_teacher_classes(g.user["user_id"])})


@handle_errors
def get_class_detail(class_id):
    return jsonify({"success": True, "data": svc.get_class_detail(class_id)})


@handle_errors
def add_class():
    class_name = (request.json or {}).get("class_name", "").strip()
    if not class_name:
        return jsonify({"success": False, "message": "Tên lớp không được trống"}), 400
    svc.add_class(g.user["user_id"], class_name)
    return jsonify({"success": True, "message": f'Tạo lớp "{class_name}" thành công'})


@handle_errors
def update_class(class_id):
    new_name = (request.json or {}).get("className", "").strip()
    if not new_name:
        return jsonify({"success": False, "message": "Tên lớp không được trống"}), 400
    svc.update_class(class_id, g.user["user_id"], new_name)
    return jsonify({"success": True, "message": "Cập nhật thành công"})


@handle_errors
def delete_class(class_id):
    svc.delete_class(class_id, g.user["user_id"])
    return jsonify({"success": True, "message": "Đã xóa lớp"})


@handle_errors
def get_class_exams(class_id):
    return jsonify({"success": True, "data": svc.get_class_exams(class_id, g.user["user_id"])})


@handle_errors
def assign_exam(class_id):
    data       = request.get_json() or {}
    exam_id    = data.get("exam_id")
    time_limit = data.get("time_limit")
    deadline   = _parse_dt(data.get("deadline"))
    open_time  = _parse_dt(data.get("open_time"))
    if not exam_id:
        return jsonify({"success": False, "message": "Thiếu exam_id"}), 400
    if not time_limit or not deadline or not open_time:
        return jsonify({"success": False, "message": "Thời gian làm bài, thời gian mở đề và hạn nộp bài không được để trống"}), 400
    try:
        time_limit = int(time_limit)
        if time_limit <= 0:
            raise ValueError
    except (ValueError, TypeError):
        return jsonify({"success": False, "message": "Thời gian làm bài không hợp lệ"}), 400
    svc.assign_exam(class_id, g.user["user_id"], exam_id, deadline, open_time, time_limit)
    return jsonify({"success": True, "message": "Đã giao bài cho lớp"})


@handle_errors
def update_exam_assignment(class_id, exam_id):
    data       = request.get_json() or {}
    time_limit = data.get("time_limit")
    deadline   = _parse_dt(data.get("deadline"))
    open_time  = _parse_dt(data.get("open_time"))
    if not time_limit or not deadline or not open_time:
        return jsonify({"success": False, "message": "Thời gian làm bài, thời gian mở đề và hạn nộp bài không được để trống"}), 400
    try:
        time_limit = int(time_limit)
        if time_limit <= 0:
            raise ValueError
    except (ValueError, TypeError):
        return jsonify({"success": False, "message": "Thời gian làm bài không hợp lệ"}), 400
    svc.update_assignment(class_id, g.user["user_id"], exam_id, deadline, open_time, time_limit)
    return jsonify({"success": True, "message": "Đã cập nhật hạn nộp"})


@handle_errors
def unassign_exam(class_id, exam_id):
    svc.unassign_exam(class_id, g.user["user_id"], exam_id)
    return jsonify({"success": True, "message": "Đã thu hồi bài thi"})


@handle_errors
def get_students_with_scores(class_id):
    return jsonify({"success": True, "data": svc.get_students_with_scores(class_id)})


@handle_errors
def get_announcements(class_id):
    return jsonify({"success": True, "data": svc.get_announcements(class_id)})


@handle_errors
def create_announcement(class_id):
    data    = request.get_json() or {}
    title   = data.get("title", "").strip()
    content = data.get("content", "").strip()
    if not title or not content:
        return jsonify({"success": False, "message": "Tiêu đề và nội dung không được trống"}), 400
    new_id = svc.create_announcement(class_id, g.user["user_id"], title, content)
    return jsonify({"success": True, "id": new_id}), 201


@handle_errors
def delete_announcement(ann_id):
    svc.delete_announcement(ann_id, g.user["user_id"])
    return jsonify({"success": True, "message": "Đã xóa thông báo"})


@handle_errors
def export_students(class_id):
    import openpyxl
    from io import BytesIO
    from urllib.parse import quote
    from flask import Response
    class_name, students = svc.get_students_for_export(class_id, g.user["user_id"])
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Danh sách học sinh"
    ws.append(["STT", "Họ và tên", "Username", "Ngày sinh", "Điểm TB", "Số bài đã làm"])
    for i, s in enumerate(students, 1):
        ws.append([i, s["full_name"], s["username"], str(s["dob"]) if s.get("dob") else "",
                   float(s["avg_score"] or 0), int(s["total_exams"] or 0)])
    buf = BytesIO()
    wb.save(buf)
    buf.seek(0)
    return Response(
        buf.getvalue(),
        mimetype="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename*=UTF-8''{quote(class_name)}.xlsx"},
    )

File: backend/src/controllers/class.controller.js
import XLSX from 'xlsx';
import * as svc from '../services/class.service.js';
import { formatDate } from '../utils/date.utils.js';
import { success, created, successMsg } from '../utils/response.js';

export const getTeacherClasses = async (req, res) => {
  const page = parseInt(req.query.page) || 1;
  const limit = parseInt(req.query.limit) || 12;
  success(res, await svc.getTeacherClasses(req.user.id, page, limit));
};

export const getClassDetail = async (req, res) => {
  const studentPage = parseInt(req.query.studentPage) || 1;
  const studentLimit = parseInt(req.query.studentLimit) || 15;
  success(res, await svc.getClassDetail(parseInt(req.params.id), studentPage, studentLimit));
};

export const addClass = async (req, res) => {
  const { className } = req.body;
  await svc.addClass(req.user.id, className.trim());
  successMsg(res, `Tạo lớp "${className.trim()}" thành công`);
};

export const updateClass = async (req, res) => {
  const { className } = req.body;
  await svc.updateClass(parseInt(req.params.id), req.user.id, className.trim());
  successMsg(res, 'Cập nhật thành công');
};

export const deleteClass = async (req, res) => {
  await svc.deleteClass(parseInt(req.params.id), req.user.id);
  successMsg(res, 'Đã xóa lớp');
};

export const getClassExams = async (req, res) => {
  success(res, await svc.getClassExams(parseInt(req.params.id), req.user.id));
};

export const assignExam = async (req, res) => {
  const { examId, timeLimit, deadline, openTime } = req.body;
  await svc.assignExam(parseInt(req.params.id), req.user.id, examId, deadline, openTime, timeLimit);
  successMsg(res, 'Đã giao bài cho lớp');
};

export const updateExamAssignment = async (req, res) => {
  const { timeLimit, deadline, openTime } = req.body;
  await svc.updateAssignment(parseInt(req.params.id), req.user.id, parseInt(req.params.examId), deadline, openTime, timeLimit);
  successMsg(res, 'Đã cập nhật hạn nộp');
};

export const unassignExam = async (req, res) => {
  await svc.unassignExam(parseInt(req.params.id), req.user.id, parseInt(req.params.examId));
  successMsg(res, 'Đã thu hồi bài tập');
};

export const getStudentsWithScores = async (req, res) => {
  success(res, await svc.getStudentsWithScores(parseInt(req.params.id)));
};

export const getAnnouncements = async (req, res) => {
  success(res, await svc.getAnnouncements(parseInt(req.params.id)));
};

export const createAnnouncement = async (req, res) => {
  const { title, content } = req.body;
  const id = await svc.createAnnouncement(parseInt(req.params.id), req.user.id, title.trim(), content.trim());
  created(res, { id });
};

export const deleteAnnouncement = async (req, res) => {
  await svc.deleteAnnouncement(parseInt(req.params.annId), req.user.id);
  successMsg(res, 'Đã xóa thông báo');
};

export const exportStudents = async (req, res) => {
  const { className, students } = await svc.getStudentsForExport(parseInt(req.params.id), req.user.id);
  const wsData = [['STT', 'Họ và tên', 'Username', 'Ngày sinh', 'Điểm TB', 'Số bài đã làm']];
  students.forEach((s, i) => {
    wsData.push([i + 1, s.full_name, s.username,
      formatDate(s.dob) ?? '',
      parseFloat(s.avg_score || 0), parseInt(s.total_exams || 0)]);
  });
  const wb = XLSX.utils.book_new();
  const ws = XLSX.utils.aoa_to_sheet(wsData);
  XLSX.utils.book_append_sheet(wb, ws, 'Danh sách học sinh');
  const buffer = XLSX.write(wb, { type: 'buffer', bookType: 'xlsx' });
  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
  res.setHeader('Content-Disposition', `attachment; filename="${encodeURIComponent(className)}.xlsx"`);
  res.send(buffer);
};
